### PR TITLE
Improve performance when posting credit memos by skipping unnecessary lookups on empty keys

### PR DESCRIPTION
**Pull request by JakovljevicDusan**
## What & why
Skip unnecessary lookups in Purch. Cr. Memo Line and Sales Cr.Memo Line when the source keys are empty:

- PurchCrMemoLine.Table.al: Only run the PurchInvLine lookup by Order No./Order Line No. when "Order No." is not empty (avoids a full-key scan when no order is referenced).
- SalesCrMemoLine.Table.al: Exit early when the credit memo header's "Applies-to Doc. No." is empty, avoiding a SalesInvoiceLine filter/find with no meaningful criteria.

Both guards prevent broad table scans during credit memo posting and related lookups, improving performance without changing functional behavior.

## Linked work

Fixes [AB#645247](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/645247)

### CODE REVIEW

File: src/Layers/ES/BaseApp/Purchases/History/PurchCrMemoLine.Table.al [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// ------------------------------------------------------------------------------------------------
+﻿// ------------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (JakovljevicDusan) [reply src/Layers/ES/BaseApp/Purchases/History/PurchCrMemoLine.Table.al]: Fix applied

File: src/Layers/ES/BaseApp/Sales/History/SalesCrMemoLine.Table.al [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// ------------------------------------------------------------------------------------------------
+﻿// ------------------------------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (JakovljevicDusan) [reply src/Layers/ES/BaseApp/Sales/History/SalesCrMemoLine.Table.al]: Fix applied

**Comment by alexei-dobriansky** [comment]
## Agentic PR Review - Round 1

# **Recommendation: Accept with Suggestions**

<details>
<summary><strong>What this PR does</strong></summary>

This PR adds two guards that skip lookups on empty keys during credit memo processing. In `Purch. Cr. Memo Line`.`GetPurchaseInvoiceLine`, the fallback lookup by "Order No."/"Order Line No." now runs only when "Order No." is not empty. In `Sales Cr.Memo Line`, the procedure exits early when the credit memo header "Applies-to Doc. No." is empty. Both are behavior-preserving. The only caller of `GetPurchaseInvoiceLine` already ignores an unfound line (`if PurchInvLine."Line No." <> 0`), and `PurchaseLine.Get(Order, '', 0)` would fail for an empty order, so nothing downstream changes. On the Sales side "Document No." is a primary key and is never empty, so the skipped `SetRange("Document No.", '')` scan never returned rows. The guards are applied consistently across all country layers.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** - *Add a regression test for the empty-key path*
The Purchase guard is safe only because the caller checks `PurchInvLine."Line No." <> 0` and `PurchaseLine.Get` fails for an empty order. This safety is not obvious from the changed line alone. Add a small test that corrects a credit memo not linked to a purchase order and confirms the order-update path is skipped, so this behavior stays locked if the caller changes later.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** These are posting-adjacent lookups in `Purch. Cr. Memo Line` and `Sales Cr.Memo Line`, but each guard only short-circuits a scan whose key is provably empty, so the returned result is identical. Risk is low. The change touches every country layer, which is correct and keeps the layers consistent. No public API or event signature changes.

**Necessity:** The change is a reasonable, low-cost performance improvement (AB#645247): it avoids a full-key table scan on `Purch. Inv. Line` and `Sales Invoice Line` when there is no order or applied invoice to look up. The scope is right for a performance fix.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=1 system=github pr=10011 round=1 by=alexei-dobriansky at=2026-08-06T09:02:33Z lastSha=9716cb1ca06fdd74577c1ebf23ed312ee73f221c reviewKey=add1a3334575ba010bca748975d3c4dbb7fb39b63b508b87b7e3190b298d8472 suggestions=S1@3a8dfdec

**Review by alexei-dobriansky — APPROVED** [review]

**Review by PredragMaricic — APPROVED** [review]
The blank-key guards are logically safe, avoid unnecessary fallback lookups, and are applied consistently across localization layers.
